In [1]:
import graphlab

In [2]:
image_train = graphlab.SFrame('image_train_data/')
image_test = graphlab.SFrame('image_test_data/')
graphlab.canvas.set_target('ipynb')

[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1537710875.log


This non-commercial license of GraphLab Create for academic use is assigned to r38411@yahoo.com and will expire on August 31, 2019.


In [3]:
image_train.head()

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 10

Data:
+-----+----------------------+------------+-------------------------------+
|  id |        image         |   label    |         deep_features         |
+-----+----------------------+------------+-------------------------------+
|  24 | Height: 32 Width: 32 |    bird    | [0.242871761322, 1.0954537... |
|  33 | Height: 32 Width: 32 |    cat     | [0.525087952614, 0.0, 0.0,... |
|  36 | Height: 32 Width: 32 |    cat     | [0.566015958786, 0.0, 0.0,... |
|  70 | Height: 32 Width: 32 |    dog     | [1.12979578972, 0.0, 0.0, ... |
|  90 | Height: 32 Width: 32 |    bird    | [1.71786928177, 0.0, 0.0, ... |
|  97 | Height: 32 Width: 32 | automobile | [1.57818555832, 0.0, 0.0, ... |
| 107 | Height: 32 Width: 32 |    dog     | [0.0, 0.0, 0.220677852631,... |
| 121 | Height: 32 Width: 32 |    bird    | [0.0, 0.23753464222, 0.0, ... |
| 136 | Height: 32 Width: 32 | automobile | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 138 | Height: 32 Width: 32 |    bird    | [0.658935725689, 0.0, 0.0,... |
+-----+----------------------+------------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [73.0, 77.0, 58.0, 71.0, 6... |
| [7.0, 5.0, 8.0, 7.0, 5.0, ... |
| [169.0, 122.0, 65.0, 131.0... |
| [154.0, 179.0, 152.0, 159.... |
| [216.0, 195.0, 180.0, 201.... |
| [33.0, 44.0, 27.0, 29.0, 4... |
| [97.0, 51.0, 31.0, 104.0, ... |
| [93.0, 96.0, 88.0, 102.0, ... |
| [35.0, 59.0, 53.0, 36.0, 5... |
| [205.0, 193.0, 195.0, 200.... |
+-------------------------------+
[10 rows x 5 columns]

In [4]:
image_train['label'].sketch_summary()


+------------------+-------+----------+
|       item       | value | is exact |
+------------------+-------+----------+
|      Length      |  2005 |   Yes    |
| # Missing Values |   0   |   Yes    |
| # unique values  |   4   |    No    |
+------------------+-------+----------+

Most frequent items:
+-------+------------+-----+-----+------+
| value | automobile | cat | dog | bird |
+-------+------------+-----+-----+------+
| count |    509     | 509 | 509 | 478  |
+-------+------------+-----+-----+------+


# Question 1: there are 4 unique labels: automobile, cat, dog and bird in image_train

## Question 2: split the data into 4 groups based on the label

In [5]:
data_list = {}
data_model = {}
i_label = image_train['label'].unique()
for i in range(len(i_label)):
    data_list[i_label[i]] = image_train.filter_by(i_label[i], 'label')
    data_model[i_label[i]] = graphlab.nearest_neighbors.create(data_list[i_label[i]],
                                                                features = ['deep_features'],
                                                                label = 'id')

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

In [6]:
the_cat = image_test[0:1]
the_cat['image'].show()

In [13]:
q2_nearest = {}
q2_nearest_1 = {}
for i in range(len(i_label)):
    q2_nearest[i_label[i]] = data_model[i_label[i]].query(the_cat)
    q2_nearest_1[i_label[i]] = q2_nearest[i_label[i]][0]['reference_label']
q2_nearest_1

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.209205    | 8.893ms      |

| Done         |         | 100         | 64.702ms     |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 12.963ms     |

| Done         |         | 100         | 63.831ms     |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 16.089ms     |

| Done         |         | 100         | 69.632ms     |

+--------------+---------+-------------+--------------+

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 9.388ms      |

| Done         |         | 100         | 66.84ms      |

+--------------+---------+-------------+--------------+

{'automobile': 49314, 'bird': 49573, 'cat': 16289, 'dog': 16976}

# Question 2 answer: nearest neighbor of the the_cat: 
# automobile model: 49314, bird_model: 49573, cat_model: 16289, dog_model: 16976

## Question 3

In [16]:
q2_nearest['automobile']

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      49314      | 39.6710582792 |  1   |
|      0      |      12389      | 40.2107393099 |  2   |
|      0      |      15299      | 41.2103305258 |  3   |
|      0      |      40118      | 41.3292164653 |  4   |
|      0      |      17111      |  41.352687775 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [22]:
for i in q2_nearest.keys():
    print i, ': ', 'avg distance = ', sum(q2_nearest[i]['distance'])/len(q2_nearest[i]['distance'])

automobile :  avg distance =  40.754806471
cat :  avg distance =  36.1557307098
bird :  avg distance =  39.2939524591
dog :  avg distance =  37.7707113618


# Question 3: avg distance from each model: 
# automobile: 40.7548, cat: 36.155, bird: 39.2939, dog: 37.77

## Question 4

In [25]:
## split the image test data based on the label

In [34]:
test_data_group = {}
for i in range(len(i_label)):
    test_data_group[i_label[i]] = image_test.filter_by(i_label[i], 'label')

In [ ]:
dog_neighbors = {}
test_group_name = 'dog'
for i in range(len(i_label)):
    dog_neighbors[i_label[i]] = data_model[i_label[i]].query(test_data_group[test_group_name], k=1)
    print i_label[i]
    print dog_neighbors[i_label[i]].head()

In [36]:
## build the SFrame

In [43]:
dog_distance = graphlab.SFrame()
for i in range(len(i_label)):
    col_name = test_group_name + '-' + i_label[i]
    dog_distance[col_name] = dog_neighbors[i_label[i]]['distance']
dog_distance.head()

Columns:
	dog-bird	float
	dog-dog	float
	dog-cat	float
	dog-automobile	float

Rows: 10

Data:
+---------------+---------------+---------------+----------------+
|    dog-bird   |    dog-dog    |    dog-cat    | dog-automobile |
+---------------+---------------+---------------+----------------+
| 41.7538647304 | 33.4773590373 | 36.4196077068 | 41.9579761457  |
| 41.3382958925 | 32.8458495684 | 38.8353268874 | 46.0021331807  |
| 38.6157590853 | 35.0397073189 | 36.9763410854 | 42.9462290692  |
| 37.0892269954 | 33.9010327697 | 34.5750072914 | 41.6866060048  |
|  38.272288694 | 37.4849250909 |  34.778824791 | 39.2269664935  |
| 39.1462089236 |  34.945165344 | 35.1171578292 | 40.5845117698  |
|  40.523040106 | 39.0957278345 | 40.6095830913 | 45.1067352961  |
| 38.1947918393 | 37.7696131032 | 39.9036867306 | 41.3221140974  |
| 40.1567131661 | 35.1089144603 | 38.0674700168 | 41.8244654995  |
| 45.5597962603 | 43.2422832585 | 42.7258732951 | 45.4976929401  |
+---------------+---------------+---------------+----------------+
[10 rows x 4 columns]

In [45]:
## now compute how many of them dog-dog is min distance

In [52]:
for i in range(len(dog_distance)):
    ans[i] = min(dog_distance[i])
ans.head()

NameError: name 'ans' is not defined